File: mod/src/main/java/org/academy/api/client/config/SkillSettingsRegistry.java
package org.academy.api.client.config;

import net.minecraft.resources.Identifier;
import org.academy.api.common.ability.Skill;
import org.jspecify.annotations.Nullable;

import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.*;

public final class SkillSettingsRegistry {
    private static final Map<Identifier, Map<String, Module>> MODULES_BY_SKILL = new LinkedHashMap<>();

    private SkillSettingsRegistry() {
    }

    public static void register(Skill skill, Module module) {
        register(skill.getKey(), module);
    }

    public static void register(Identifier skillId, Module module) {
        var modules = MODULES_BY_SKILL.computeIfAbsent(skillId, ignored -> new LinkedHashMap<>());
        if (modules.putIfAbsent(module.id(), module) != null) {
            throw new IllegalStateException(
                    "Skill settings module '" + module.id() + "' is already registered for " + skillId);
        }
    }

    public static void unregister(Identifier skillId, String moduleId) {
        var modules = MODULES_BY_SKILL.get(skillId);
        if (modules == null) return;
        modules.remove(moduleId);
        if (modules.isEmpty()) MODULES_BY_SKILL.remove(skillId);
    }

    public static List<Module> getModules(Skill skill) {
        var modules = MODULES_BY_SKILL.get(skill.getKey());
        return modules == null ? List.of() : List.copyOf(modules.values());
    }

    public record Module(String id, String titleKey, List<Entry> entries) {
        public Module {
            if (id.isBlank()) {
                throw new IllegalArgumentException("Skill settings module id cannot be blank");
            }
            var entryIds = new HashSet<String>();
            for (var entry : entries) {
                if (!entryIds.add(entry.id())) {
                    throw new IllegalArgumentException(
                            "Skill settings module '" + id + "' contains duplicate entry ids");
                }
            }
        }
    }

    public sealed interface Entry permits Toggle, IntegerRange, Choice, FloatRange, Action {
        String id();

        String labelKey();
    }

    public record Toggle(String id, String labelKey, BooleanSupplier getter,
                         Consumer<Boolean> setter) implements Entry {
    }

    public record IntegerRange(String id, String labelKey, int min, int max, int step, IntSupplier getter,
                               IntConsumer setter) implements Entry {
        public IntegerRange {
            if (min > max) {
                throw new IllegalArgumentException("Integer setting '" + id + "' has an invalid range");
            }
            if (step <= 0) {
                throw new IllegalArgumentException("Integer setting '" + id + "' must use a positive step");
            }
        }
    }

    public record Choice(String id, String labelKey, List<String> optionKeys, IntSupplier getter, IntConsumer setter,
                         BooleanSupplier available, String unavailableKey) implements Entry {
        public Choice {
            if (optionKeys.isEmpty()) {
                throw new IllegalArgumentException("Choice setting '" + id + "' must have at least one option");
            }
            if (optionKeys.stream().anyMatch(String::isBlank)) {
                throw new IllegalArgumentException(
                        "Choice setting '" + id + "' contains an empty option translation key");
            }
        }

        public int clampIndex(int index) {
            return Math.clamp(index, 0, optionKeys.size() - 1);
        }
    }

    @FunctionalInterface
    public interface FloatSupplier {
        float getAsFloat();
    }

    public record FloatRange(String id, String labelKey, float min, float max, float step, FloatSupplier getter,
                             Consumer<Float> setter, Runnable commit,
                             @Nullable Function<Float, String> formatter) implements Entry {
        public FloatRange(
                String id,
                String labelKey,
                float min,
                float max,
                float step,
                FloatSupplier getter,
                Consumer<Float> setter,
                Runnable commit
        ) {
            this(id, labelKey, min, max, step, getter, setter, commit, null);
        }

        public FloatRange {
            if (!Float.isFinite(min) || !Float.isFinite(max) || min > max) {
                throw new IllegalArgumentException("Float setting '" + id + "' has an invalid range");
            }
            if (!Float.isFinite(step) || step <= 0f) {
                throw new IllegalArgumentException("Float setting '" + id + "' must use a positive finite step");
            }
        }

        public String formatValue(float value) {
            if (formatter != null) return formatter.apply(value);
            var range = max - min;
            var normalized = range > 0f ? Math.clamp((value - min) / range, 0f, 1f) : 0f;
            return Math.round(normalized * 100f) + "%";
        }

        public float quantize(float value) {
            if (!Float.isFinite(value)) return min;
            var index = Math.round((value - min) / step);
            return Math.clamp(min + index * step, min, max);
        }
    }

    public record Action(String id, String labelKey, String buttonKey, Runnable action) implements Entry {
    }
}

File: src/main/kotlin/org/academy/api/client/config/SkillSettingsRegistry.kt
package org.academy.api.client.config

import net.minecraft.resources.Identifier
import org.academy.api.common.ability.Skill
import java.util.function.BooleanSupplier
import java.util.function.Consumer
import java.util.function.IntConsumer
import java.util.function.IntSupplier
import kotlin.math.roundToInt

/**
 * Registers client-side, per-skill settings displayed by the Skill Settings app.
 * Modules keep unrelated skill options isolated and allow new settings to be added
 * without changing the app implementation.
 */
object SkillSettingsRegistry {
    private val modulesBySkill = LinkedHashMap<Identifier, LinkedHashMap<String, Module>>()

    @JvmStatic
    @Synchronized
    fun register(skill: Skill, module: Module) {
        register(skill.key, module)
    }

    @JvmStatic
    @Synchronized
    fun register(skillId: Identifier, module: Module) {
        val modules = modulesBySkill.getOrPut(skillId) { LinkedHashMap() }
        check(modules.putIfAbsent(module.id, module) == null) {
            "Skill settings module '${module.id}' is already registered for $skillId"
        }
    }

    @JvmStatic
    @Synchronized
    fun unregister(skillId: Identifier, moduleId: String) {
        modulesBySkill[skillId]?.let { modules ->
            modules.remove(moduleId)
            if (modules.isEmpty()) modulesBySkill.remove(skillId)
        }
    }

    @JvmStatic
    @Synchronized
    fun getModules(skill: Skill): List<Module> {
        return modulesBySkill[skill.key]?.values?.toList() ?: emptyList()
    }

    data class Module(
        val id: String,
        val titleKey: String,
        val entries: List<Entry>
    ) {
        init {
            require(id.isNotBlank()) { "Skill settings module id cannot be blank" }
            require(entries.map(Entry::id).distinct().size == entries.size) {
                "Skill settings module '$id' contains duplicate entry ids"
            }
        }
    }

    sealed interface Entry {
        val id: String
        val labelKey: String
    }

    data class Toggle(
        override val id: String,
        override val labelKey: String,
        val getter: BooleanSupplier,
        val setter: Consumer<Boolean>
    ) : Entry

    data class IntegerRange(
        override val id: String,
        override val labelKey: String,
        val min: Int,
        val max: Int,
        val step: Int = 1,
        val getter: IntSupplier,
        val setter: IntConsumer
    ) : Entry {
        init {
            require(min <= max) { "Integer setting '$id' has an invalid range" }
            require(step > 0) { "Integer setting '$id' must use a positive step" }
        }
    }

    fun interface FloatSupplier {
        fun getAsFloat(): Float
    }

    data class FloatRange(
        override val id: String,
        override val labelKey: String,
        val min: Float,
        val max: Float,
        val step: Float = 0.05f,
        val getter: FloatSupplier,
        val setter: Consumer<Float>,
        val commit: Runnable
    ) : Entry {
        init {
            require(min.isFinite() && max.isFinite() && min <= max) {
                "Float setting '$id' has an invalid range"
            }
            require(step.isFinite() && step > 0f) {
                "Float setting '$id' must use a positive finite step"
            }
        }

        fun quantize(value: Float): Float {
            if (!value.isFinite()) return min
            val index = ((value - min) / step).roundToInt()
            return (min + index * step).coerceIn(min, max)
        }
    }

    data class Action(
        override val id: String,
        override val labelKey: String,
        val buttonKey: String,
        val action: Runnable
    ) : Entry
}
